docs/adr/0010-mise-en-place-de-bounded-contexts.md

```diff
@@ -0,0 +1,176 @@
+1. Utilisation de **plugins Hapi.js** pour matérialiser ces modules de code
+1. **Duplication de circonstance** pour les données utiles à plusieurs contextes
+
+## Conséquences
+
+### Architecture hexagonale 
+
+Les 3 couches retenues sont : 
+- application
+- domain
+- infrastructure
+
+### Contextes métiers
+
+Les contextes métier retenus sont :
+- content : référentiel pédagogique, caching des épreuves
+- assessment : poser des questions, recueillir et évaluer des réponses
+- accounts : comptes, accès, mots de passe, interconnexion 
+- profile : calculer un profil de compétence, ainsi que ses attributes (ex : badges)
+- training : analyse, accompagnement, prescription, formation 
+- certification : examen, logisitique, jury
+- admin : back-office, CRUD
+
+> Remarque : cet ordre correspond à l'apparition historique des fonctionnalités de Pix : nous avons commencé par 
+> élaborer des questions, que nous avons converties pour pouvoir les soumettre aux utilisateurs, sur des tests de démo.
+> Nous avons ensuite ajouter les fonctionnalités relatives aux comptes & accès, avec persistence des réponses, ce qui 
+> nous a permis d'élaborer un profil de compétences pour chacun et de proposer une expérience adaptée et adaptative. 
+> Suite à ça, nous avons développé l'activité de prescription, puis de certification (dont l'outillage pour le jury).
+> Finalement, nous avons dû nous attaquer à nous outiller nous-mêmes pour pouvoir gérer l'assistance aux utilisateurs 
+> et leurs demandes de support.   
+
+Ces contextes sont susceptibles d'être modifiés en fonction des évolutions des activités, du business model et autres aléas inhérents au GIP PIX, ainsi que des obligations techniques de la plateforme.
+
+### Anatomie du code
+
+La nouvelle arborescence générale du code de Pix API devient :
+
+```
+api
+ └ bin
+    └ www
+ └ db
+ └ lib
+    └ accounts
+    └ admin
+    └ assessment
+    └ certification
+    └ content
+    └ profile
+    └ training
+ └ server.js
+```
+
+> Les répertoires ou fichiers non impactés par cet décision d'architecture ne sont pas mentionnés. 
+
+Chaque module de code respecte la même organisation de code :
+
+```
+api/lib/<context>
+ └ application
+    └ routes
+    └ controllers
+    └ prehandlers
+    └ …
+ └ domain
+    └ models (aggregates, entities, value objects)
+    └ services
+    └ usecases
+    └ validators
+    └ …
+ └ infrastructure
+    └ adapters
+    └ converters
+    └ serializers
+    └ repositories
+    └ …
+```
+
+### Utilisation de plugins Hapi.js
+
+Hapi.js recommande de découper son code en modules métier. À ces fins, le framework propose le mécanisme des plugins, l'équivalent des middlewares sous Express.js. 
+
+> hapi has an extensive and powerful plugin system that allows you to very easily break your application up into isolated pieces of business logic, and reusable utilities. You can either add an existing plugin to your application, or create your own.
+
```

Comment:
J'aimerais mieux comprendre ce que peuvent apporter les _plugins_ Hapi.js ici. A priori j'aurais dit que le choix d'utiliser des _plugins_ était un détail d'implémentation d'un des _adapters_, en quoi est-ce structurant ? J'imagine qu'un exemple aiderait à réfléchir sur le sujet.

Reply:
Si je comprend bien l'utilisation de plug-ins hapi forcent la séparation dure entre contextes.
Je suis partagé entre :
1) avoir une séparation dure pour que ça fasse mal et nous oblige à faire des choses très carré dès le début : on crée les X plug-ins hapi et on y transfère petit à petit les trucs qui sont vraiment vraiment clean (= sans dépendances vers les autres contextes)
2) avoir une séparation soft (ie de simples dossiers) pour avoir un truc relativement permissif qui nous permettre une séparation grosse maille de ce qui ira où à terme en gardant dans un premier temps les dépendances (ex. L'objet `assessment` est déplacé dans le contexte/dossier `evaluation`, les `usecases` de certification, qui eux sont dans le contexte/dossier `certification` en dependent encore "illégalement". Petit à petit, on identifie les imports "illégaux" et on refacto. Quand il n'y a plus d'imports illégaux dans un contexte, on en fait un plug-in (s'il existe une plus value à le faire))

Reply:
C'est effectivement un détail d'implémentation, mais à un momoent, on ne peut pas rester que dans l'abstrait. J'ai tenté d'apporter un éclairage plus concret (peut-être trop). L'idée c'est qu'il y ait "en sortie du module / service" (ici, au sens activité proposée) un machin plutôt isolé mais facilement pluggable dans l'architecture générale (i.e. des définition de routes et d'outils au format attendu par Hapi).